File: osmani/osmani/report/user_activity_report/user_activity_report.py
import frappe
from frappe import _


DOCTYPES = [
    ("Purchase Invoice", "purchase_invoice"),
    ("Sales Order", "sales_order"),
    ("Sales Invoice", "sales_invoice"),
    ("Payment Entry", "payment_entry"),
    ("Journal Entry", "journal_entry"),
    ("Employee Advance", "employee_advance"),
    ("Expense Claim", "expense_claim"),
]

DOCSTATUS_MAP = {"Draft": 0, "Submitted": 1, "Cancelled": 2}


def _normalize_doctype_list(value):
    if not value:
        return []
    if isinstance(value, (list, tuple)):
        return [v for v in value if v]
    # accept comma separated string
    return [v.strip() for v in str(value).split(",") if v.strip()]


def execute(filters=None):
    filters = filters or {}

    selected = set(_normalize_doctype_list(filters.get("doctype_list")))
    selected_doctypes = [pair for pair in DOCTYPES if not selected or pair[0] in selected]

    # Build columns dynamically based on selected doctypes
    columns = [
        {"label": _("User"), "fieldname": "user", "fieldtype": "Link", "options": "User", "width": 180}
    ]
    for dt, key in selected_doctypes:
        columns.append({"label": _(dt), "fieldname": key, "fieldtype": "Int", "width": 130})
    columns.append({"label": _("Total"), "fieldname": "total", "fieldtype": "Int", "width": 130})

    # Common conditions
    conditions = ["1=1"]
    params = {}

    if filters.get("from_date"):
        conditions.append("creation >= %(from_date)s")
        params["from_date"] = filters.get("from_date")

    if filters.get("to_date"):
        conditions.append("creation <= %(to_date)s")
        params["to_date"] = filters.get("to_date")

    if filters.get("user"):
        conditions.append("owner = %(user)s")
        params["user"] = filters.get("user")

    docstatus_label = filters.get("docstatus")
    if docstatus_label and docstatus_label != "All":
        conditions.append("docstatus = %(docstatus)s")
        params["docstatus"] = DOCSTATUS_MAP.get(docstatus_label, docstatus_label)

    where_clause = " AND ".join(conditions)

    # Aggregate per doctype with GROUP BY owner
    rows_by_user = {}
    for dt, key in selected_doctypes:
        table = f"`tab{dt}`"
        res = frappe.db.sql(
            f"""
            SELECT owner, COUNT(*) AS cnt
            FROM {table}
            WHERE {where_clause}
            GROUP BY owner
            """,
            params,
            as_dict=True,
        )
        for r in res:
            u = r["owner"]
            row = rows_by_user.setdefault(u, {"user": u})
            row[key] = r["cnt"]

    # Build final data, fill missing zeros, and totals
    data = []
    for u, row in rows_by_user.items():
        total = 0
        for dt, key in selected_doctypes:
            val = int(row.get(key, 0) or 0)
            row[key] = val
            total += val
        row["total"] = total
        data.append(row)

    # Sort by total desc then user asc
    data.sort(key=lambda x: (-x.get("total", 0), x.get("user", "")))

    return columns, data

File: osmani/osmani/report/user_activity_report/user_activity_report.js
frappe.query_reports["User Activity Report"] = {
    filters: [
        {
            fieldname: "from_date",
            label: "From Date",
            fieldtype: "Date"
        },
        {
            fieldname: "to_date",
            label: "To Date",
            fieldtype: "Date"
        },
        {
            fieldname: "user",
            label: "User",
            fieldtype: "Link",
            options: "User"
        },
        {
            fieldname: "doctype_list",
            label: "Document Types",
            fieldtype: "MultiSelectList",
            get_data: function() {
                const doctypes = [
                    "Purchase Invoice",
                    "Sales Order",
                    "Sales Invoice",
                    "Payment Entry",
                    "Journal Entry",
                    "Employee Advance",
                    "Expense Claim",
                ];
                return doctypes.map(d => ({
                    label: d,
                    value: d
                }));
            }
        },
        {
            fieldname: "docstatus",
            label: "Status",
            fieldtype: "Select",
            options: ["All", "Draft", "Submitted", "Cancelled"].join("\n"),
            default: "All"
        }
    ],
    onload: function(report) {
        report.page.set_title("User Activity Summary");

        // Robust Export to Excel (uses built-in if available, else falls back to API)
        const download_excel = (report_name, filters) => {
            try {
                if (frappe.query_report && typeof frappe.query_report.download === "function") {
                    frappe.query_report.download();
                    return;
                }
            } catch (e) {
                // ignore and use fallback
            }
            const url = "/api/method/frappe.desk.query_report.download" +
                `?report_name=${encodeURIComponent(report_name)}` +
                `&file_format=Excel` +
                `&filters=${encodeURIComponent(JSON.stringify(filters || {}))}`;
            window.open(url);
        };

        // report.page.add_button("Export Excel", () => {
        //     const f = report.get_values();
        //     download_excel("User Activity Report", f);
        // });

        // Event delegation for clickable counts to ensure navigation works reliably
        report.$wrapper.on("click", "a.count-link", (ev) => {
            ev.preventDefault();
            const a = ev.currentTarget;
            const doctype = a.getAttribute("data-doctype");
            const user = a.getAttribute("data-user");
            const f = report.get_values();

            let dtList = f.doctype_list;
            if (typeof dtList === "string" && dtList) {
                dtList = dtList.split(",").map(s => s.trim()).filter(Boolean);
            }
            if (!Array.isArray(dtList)) {
                dtList = dtList ? [dtList] : [];
            }
            if (!dtList.includes(doctype)) {
                dtList.push(doctype);
            }

            const route_opts = {
                from_date: f.from_date,
                to_date: f.to_date,
                user: user,
                doctype: doctype,
                docstatus: f.docstatus,
                doctype_list: dtList,
            };
            frappe.set_route("query-report", "User Activity Detail", route_opts);
        });
    },
    formatter: function(value, row, column, data, default_formatter) {
        const formatted_value = default_formatter(value, row, column, data);
        const clickables = [
            "purchase_invoice",
            "sales_order",
            "sales_invoice",
            "payment_entry",
            "journal_entry",
            "employee_advance",
            "expense_claim",
        ];
        if (column && clickables.includes(column.fieldname) && value) {
            const dtLabelMap = {
                purchase_invoice: "Purchase Invoice",
                sales_order: "Sales Order",
                sales_invoice: "Sales Invoice",
                payment_entry: "Payment Entry",
                journal_entry: "Journal Entry",
                employee_advance: "Employee Advance",
                expense_claim: "Expense Claim",
            };

            const user = data.user;
            const doctype = dtLabelMap[column.fieldname];

            return `<a href="#" data-doctype="${doctype}" data-user="${user}" class="count-link">${formatted_value}</a>`;
        }
        return formatted_value;
    },
    after_datatable_render: function(report) {
        const el = report.$wrapper[0];
        el.querySelectorAll("a.count-link").forEach(a => {
            a.addEventListener("click", (ev) => {
                ev.preventDefault();
                const doctype = a.getAttribute("data-doctype");
                const user = a.getAttribute("data-user");
                const f = report.get_values();

                // Normalize MultiSelectList value for doctypes to preserve in Detail and Back navigation
                let dtList = f.doctype_list;
                if (typeof dtList === "string" && dtList) {
                    dtList = dtList.split(",").map(s => s.trim()).filter(Boolean);
                }
                if (!Array.isArray(dtList)) {
                    dtList = dtList ? [dtList] : [];
                }
                // Ensure clicked doctype is included
                if (!dtList.includes(doctype)) {
                    dtList.push(doctype);
                }

                const route_opts = {
                    from_date: f.from_date,
                    to_date: f.to_date,
                    user: user,
                    doctype: doctype,
                    docstatus: f.docstatus,
                    doctype_list: dtList
                };
                frappe.set_route("query-report", "User Activity Detail", route_opts);
            });
        });
    }
};
